include/revng/Pipeline/Structure.h

```diff
@@ -0,0 +1,224 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <initializer_list>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/Pipeline/Errors.h"
+#include "revng/Pipeline/Step.h"
+
+namespace Pipeline {
+
+class PipelineExecutionEntry {
+public:
+  Step *ToExecute;
+  ContainersStatus Objectives;
+
+  PipelineExecutionEntry(Step &ToExecute, ContainersStatus Objectives) :
+    ToExecute(&ToExecute), Objectives(std::move(Objectives)) {}
+};
+
+class BoundedStep {
+private:
+  std::string PreviousStepName;
+  Step CurrentStep;
+
+public:
+  BoundedStep(std::string PreviousStepName, Step CurrentStep) :
+    PreviousStepName(std::move(PreviousStepName)),
+    CurrentStep(std::move(CurrentStep)) {}
+
+  bool hasPredecessor() const { return not PreviousStepName.empty(); }
+  const Step &getStep() const { return CurrentStep; }
+  Step &getStep() { return CurrentStep; }
+  const std::string &getPreviousStepName() const { return PreviousStepName; }
+
+  Step *operator->() { return &CurrentStep; }
+  const Step *operator->() const { return &CurrentStep; }
+
+  Step &operator*() { return CurrentStep; }
+  const Step &operator*() const { return CurrentStep; }
+};
+
+/// A pipeline structure is a sequence of steps connected one to
+/// the next, this is the level of abstraction where operations
+/// such as cross step requirements deduction can be asked.
+class Structure {
```

Comment:
Can we get a name more intuitive than `Structure`? To me sounds like `StepList` or `RunnableStepList`.